feat(managers) add task duration and start time

ds-skoropad opened the pull request:
Приветствую!

Выполнял задание с "горящими пятками", и все же прошелся по всему коду.
Надеюсь критичных  замечаний будет не много.
Но я как и всегда буду рад любым замечаниям, так как указание на мои ошибки для меня бесценно.

ds-skoropad committed:
feat(managers) add task duration and start time

Refactoring code using streams
Abstract classes in tests for code reduction

Sprint 8 Review 1

Ivanov-s-tmn commented on .idea/misc.xml:
Добавь директорию .idea в исключения gitignore, эта папка генерируется локально идеей и не нужна в репозитории

Ivanov-s-tmn commented on src/ru/yandex/practicum/kanban/manager/TaskManager.java:
```diff
@@ -5,44 +5,41 @@
-    boolean removeEpicTask(int id);
+    boolean removeEpic(int id);
 
     void setHistoryManager(HistoryManager historyManager);
```
А зачем сеттер для менеджера истории? Инициализируй в конструкторе менеджером по умолчанию

Ivanov-s-tmn commented on src/ru/yandex/practicum/kanban/task/Task.java:
```diff
@@ -1,25 +1,50 @@
+    protected Duration duration;
+    protected LocalDateTime startTime;
+
+    public Task() {
+        this.title = "";
+        this.description = "";
+        this.status = StatusTask.NEW;
+        this.duration = Duration.ZERO;
+        this.id = 0;
+    }
```
Вызови другой конструктор, а не вот это вот все)
как пример, `this("", "", 0, StatusTask.NEW, Duration.ZERO);`
С остальными конструкторами можно так же, чтобы не дублировать инициализацию одних и тех же полей

Ivanov-s-tmn commented on test/ru/yandex/practicum/kanban/manager/InMemoryHistoryManagerTest.java:
```diff
@@ -2,82 +2,71 @@
-    private Task task1;
-    private Task task2;
-    private Task task3;
+class InMemoryHistoryManagerTest extends HistoryManagerTest<HistoryManager> {
```
Для менеджера можно и не делать абстрактный параметризованный класс, у нас одна реализация и мы ничего не дублируем

Ivanov-s-tmn commented on test/ru/yandex/practicum/kanban/manager/InMemoryTaskManagerTest.java:
```diff
@@ -1,426 +1,252 @@
+    }
 
-        // Эпик2
-        sub1InEpic2 = getCopySubTaskWithNewStatus(sub1InEpic2, StatusTask.DONE);
-        taskManager.updateTask(sub1InEpic2);
-        sub2InEpic2 = getCopySubTaskWithNewStatus(sub2InEpic2, StatusTask.DONE);
-        taskManager.updateTask(sub2InEpic2);
-        assertNotEquals(StatusTask.NEW, epic2.getStatus(), StatusTask.NEW + " status not expected");
+    @Test
+    @Override
+    void shouldBeExceptionValidateTimeForUpdateOverlayTask() {
+        super.shouldBeExceptionValidateTimeForUpdateOverlayTask();
+    }
 
-        sub1InEpic2 = getCopySubTaskWithNewStatus(sub1InEpic2, StatusTask.NEW);
-        taskManager.updateTask(sub1InEpic2);
-        assertNotEquals(StatusTask.NEW, epic2.getStatus(), StatusTask.NEW + " status not expected");
+    @Test
+    @Override
+    void shouldBeExceptionValidateTimeForUpdateOverlaySub() {
+        super.shouldBeExceptionValidateTimeForUpdateOverlaySub();
+    }
 
-        sub2InEpic2 = getCopySubTaskWithNewStatus(sub2InEpic2, StatusTask.NEW);
-        taskManager.updateTask(sub2InEpic2);
-        assertEquals(StatusTask.NEW, epic2.getStatus(), StatusTask.NEW + " status expected");
+    @Test
+    @Override
+    void notShouldBeExceptionValidateTimeForUpdateOverlayEpic() {
+        super.notShouldBeExceptionValidateTimeForUpdateOverlayEpic();
+    }
 
-        sub2InEpic2 = getCopySubTaskWithNewStatus(sub2InEpic2, StatusTask.DONE);
-        taskManager.updateTask(sub2InEpic2);
-        taskManager.removeSubTask(sub2InEpic2.getId());
-        assertEquals(StatusTask.NEW, epic2.getStatus(), StatusTask.NEW + " status expected");
+    @Test
+    @Override
+    void shouldBeExceptionOutOfRangeLeftIntervals() {
+        super.shouldBeExceptionOutOfRangeLeftIntervals();
```
Тут не нужно переопределять тестовые методы, просто пометь их аннотацией `@Test` в `TaskManagerTest` и все.
В тестах для файлового менеджера аналогично

Ivanov-s-tmn commented on src/ru/yandex/practicum/kanban/manager/InMemoryTaskManager.java:
```diff
@@ -1,283 +1,415 @@
+    public Optional<EpicTask> getEpic(int id) {
+        Optional<EpicTask> epicTask = Optional.ofNullable(epicGroup.get(id));
 
-        if (historyManager != null) {
-            historyManager.add(epicTask);
+        if (historyManager != null && epicTask.isPresent()) {
+            historyManager.add(epicTask.get());
```
Сомнительная реализация через Optional, как по мне было проще и лучше)

`historyManager != null` тоже не нужна, по идее у нас таск менеджер всегда должен иметь менеджер истории, его просто всегда можно инициализировать через `Managers.getDefaultHistory()`

Ivanov-s-tmn commented on src/ru/yandex/practicum/kanban/manager/InMemoryTaskManager.java:
```diff
@@ -1,283 +1,415 @@
+
+    // Метод пересечения двух задач (решение до интервального планирования)
+    private boolean overlayTasks(Task task1, Task task2) {
+        return task1 != null && task2 != null && task1.getStartTime().isPresent() && task2.getStartTime().isPresent() &&
+                task1.getEndTime().isPresent() && task2.getEndTime().isPresent() &&
+                task1.getStartTime().get().isBefore(task2.getEndTime().get()) &&
+                task2.getStartTime().get().isBefore(task1.getEndTime().get());
```
Получилось сложновато, тут можно написать более короткое условие:

Самое простое - это написать условие для непересекающихся интервалов.

Непересечение, это когда начало одного интервала больше конца другого интервала, или конец одного интервала меньше начала другого:
a.start > b.end OR a.end < b.start

Соответственно, пересечение - это отрицание того, что мы получили - т.е. если для интервалов выполняется NOT( a.start > b.end OR a.end < b.start )

Теперь раскрываем по правилам логики скобки и получаем a.start <= b.end AND a.end >= b.start

Ivanov-s-tmn commented on src/ru/yandex/practicum/kanban/manager/InMemoryTaskManager.java:
```diff
@@ -1,283 +1,415 @@
+                }
+                task.setId(nextId);
+                taskGroup.put(nextId, task);
+                if (task.getStartTime().isPresent()) {
```
А не проще делать проверку на наличие времени в самом методе `overlayPrioritizedTasksInterval()`?

А проверять на пересечение непосредственно перед добавлением в отсортированный список, так как в ТЗ есть пункт, что у задач может быть не задано время, соответственно такие задачи должны быть в хешмапе, но их не будет в отсортированном списке

Ivanov-s-tmn commented on src/ru/yandex/practicum/kanban/manager/InMemoryTaskManager.java:
```diff
@@ -1,283 +1,415 @@
-                epicTask.setId(id);
-                epicTaskGroup.put(id, epicTask);
+                EpicTask epic = (EpicTask) task;
+                epic.setId(nextId);
+                epic.setStatus(StatusTask.NEW);
+                epic.setSubIds(new ArrayList<>());
+                epic.setDuration(Duration.ZERO);
+                epic.setStartTime(null);
```
Создай через конструктор?

ds-skoropad committed:
delete .idea path

ds-skoropad committed:
refactor(task) rebuild the constructors

For optimal creation of tasks in the project.
And also:
* The history manager is now available by default
* "Test-descendant" without the unnecessary @Override
Some extra fix.

Sprint 8 Review 2

Ivanov-s-tmn commented on src/ru/yandex/practicum/kanban/manager/FileBackedTaskManager.java:
При загрузке из файла (восстановление менеджера) нужно предусмотреть добавление задач/подзадач в отсортированный список

ds-skoropad replied:
Это провал :)

Ivanov-s-tmn commented on test/ru/yandex/practicum/kanban/manager/FileBackedTaskManagerTest.java:
В тестовых классах `FileBackedTaskManagerTest` и `InMemoryTaskManagerTest`  нет необходимости переопределять тестовые методы `TaskManagerTest`, если пометить их аннотацией `@Test`, они будут автоматически запускаться в наследниках

ds-skoropad committed:
test(managers): do not override test methods

Sprint 8 Review 3

ds-skoropad committed:
feat(managers): add priority list to file support

Sprint 8 Review 3
